### ISSUE TITLE
UBSan: unsigned-integer-overflow false positive that should be handled better

### ISSUE THREAD

**Issue by luke-jr** [llvm/llvm-project#57794]
```c++
std::vector<char> whatever;
for (auto i{whatever.size()}; i--; ) {
    // do something
}
```

This seems to be the most logically optimal way to go over all elements of a vector (or similar). Yet unsigned-integer-overflow has a false positive on it. I would think it'd be pretty simple to detect that the overflow results in `i` going out of scope before any access?

**Comment by firewave** [comment]
I came across something similar a few years ago in the beast library: https://github.com/boostorg/beast/issues/1342. I no longer use the code but might possibly be able to pull the examples out of the code.

**Comment by nikic** [comment]
Note that the unsigned integer overflow sanitizer is not enabled by default and does **not** indicate undefined behavior. People should **not** be enabling it unless they have a very specialized use case, for example to provide additional signal for coverage-driven fuzzing (in which case it is used solely for the additional control flow edges, the actual reports get suppressed).

**Comment by firewave** [comment]
> Note that the unsigned integer overflow sanitizer is not enabled by default and does **not** indicate undefined behavior

I am aware of that (and it is also stated in the documentation). But that is also true for other checks.

On a side note... it's also a common argument ("it's not undefined behavior") if people don't like or understand some of the warnings. 😐

> People should **not** be enabling it unless they have a very specialized use case, for example to provide additional signal for coverage-driven fuzzing (in which case it is used solely for the additional control flow edges, the actual reports get suppressed).

That's very interesting.

The documentation sounds quite different though:
```
Unlike signed integer overflow, this is not undefined behavior, but it is often unintentional.
```
I think what you have written should be put there instead.

**Comment by luke-jr** [comment]
It's still undesirable to have a false positive when it can be detected as such. (Possibly also indicates missing an optimisation: there's no need to actually increment in this case either)

**Comment by nikic** [comment]
It's a bit hard to talk about false positives for a sanitizer that is not backed by the C++ spec, but if we reasonably assume that unsigned-integer-overflow should follow exactly the same semantics as signed-integer-overflow, then this is not a false positive (see https://clang.godbolt.org/z/z7ebGG7M1 for comparison). This is because overflow in C++ is immediate undefined behavior, whether the result gets used is irrelevant.

**Comment by luke-jr** [comment]
That's not a reasonable assumption.

Signed integer overflow is UB, even if the result is discarded.

There's no reason anyone would want a warning in this scenario.
